### PR TITLE
PS-4892 : Re-implement expanded fast index creation in 8.0 DD

### PR DESCRIPTION
**Pull request by satya-bodapati**
Expanded fast index creation is reimplemented in 8.0.

The strategy remains same as 5.7 i.e. for ALTER ALGORITHM=COPY,

1. On the intermediate table, remove secondary indexes
2. copy data
3. Create secondary indexes.

Specifically 8.0 challenges are with DD to handle the
intermediate DD table object with removed secondary indexes.

To handle this, an in-memory flag is set in DD objects. Note that
this intermediate table is not accessible by other threads or statements.
So it is safe.

In InnoDB, dd checks for the disabled for secondary indexes which are marked
disabled. When the secondary indexes are restored, full checks are done.

Apart from that, for temporary tables, there is intermediate full commit
after copy stage.

Since inplace alter APIs do not commit the transaction in 8.0, we have trx
in active state. Because of the full intermediate commit in between, future
attempts to commit will not succeed.

It looks safe to disable the intermediate full commit for all temp tables but
we play safe and disable only when expanded fast index creation is enabled.

This commit reverts PS-5195 and PS-5124 (they are about disabling
expanded fast index creation until it is reimplemented).

### CODE REVIEW

File: sql/dd/impl/types/index_impl.h [hunk 1]
```diff
@@ -121,6 +121,14 @@ class Index_impl : public Entity_object_impl, public Index {
 
   virtual void set_hidden(bool hidden) { m_hidden = hidden; }
 
+  /////////////////////////////////////////////////////////////////////////
+  // is_disabled.
+  /////////////////////////////////////////////////////////////////////////
+
+  virtual bool is_disabled() const { return m_disabled; }
```
Comment (laurynas-biveinis) [inline sql/dd/impl/types/index_impl.h]:
```suggestion
  virtual bool is_disabled() const noexcept { return m_disabled; }
```

File: sql/dd/impl/types/index_impl.h [hunk 2]
```diff
@@ -121,6 +121,14 @@ class Index_impl : public Entity_object_impl, public Index {
...
+
+  virtual void set_disabled(bool disable) { m_disabled = disable; }
```
Comment (laurynas-biveinis) [inline sql/dd/impl/types/index_impl.h]:
```suggestion
  virtual void set_disabled(bool disable) noexcept { m_disabled = disable; }
```

File: sql/dd/impl/types/partition_index_impl.h [hunk 3]
```diff
@@ -73,6 +73,16 @@ class Partition_index_impl : public Weak_object_impl, public Partition_index {
 
   virtual bool store_attributes(Raw_record *r);
 
+  /////////////////////////////////////////////////////////////////////////
+  // is_disabled.
+  /////////////////////////////////////////////////////////////////////////
+
+  virtual bool is_disabled() const { return m_index->is_disabled(); }
```
Comment (laurynas-biveinis) [inline sql/dd/impl/types/partition_index_impl.h]:
```suggestion
  virtual bool is_disabled() const noexcept { return m_index->is_disabled(); }
```

File: sql/dd/impl/types/partition_index_impl.h [hunk 4]
```diff
@@ -73,6 +73,16 @@ class Partition_index_impl : public Weak_object_impl, public Partition_index {
...
+
+  virtual void set_disabled(bool disable) { m_index->set_disabled(disable); }
```
Comment (laurynas-biveinis) [inline sql/dd/impl/types/partition_index_impl.h]:
```suggestion
  virtual void set_disabled(bool disable) noexcept { m_index->set_disabled(disable); }
```

File: sql/dd/impl/types/partition_index_impl.h [hunk 5]
```diff
@@ -73,6 +73,16 @@ class Partition_index_impl : public Weak_object_impl, public Partition_index {
...
+
+  virtual Index::enum_index_type type() const { return m_index->type(); }
```
Comment (laurynas-biveinis) [inline sql/dd/impl/types/partition_index_impl.h]:
```suggestion
  virtual Index::enum_index_type type() const noexcept { return m_index->type(); }
```

File: sql/dd/types/index.h [hunk 6]
```diff
@@ -100,6 +100,13 @@ class Index : virtual public Entity_object {
   virtual bool is_hidden() const = 0;
   virtual void set_hidden(bool hidden) = 0;
 
+  /////////////////////////////////////////////////////////////////////////
+  // disabled.
+  /////////////////////////////////////////////////////////////////////////
+
+  virtual bool is_disabled() const = 0;
```
Comment (laurynas-biveinis) [inline sql/dd/types/index.h]:
```suggestion
  virtual bool is_disabled() const noexcept = 0;
```

File: sql/dd/types/index.h [hunk 7]
```diff
@@ -100,6 +100,13 @@ class Index : virtual public Entity_object {
...
+  virtual void set_disabled(bool disable) = 0;
```
Comment (laurynas-biveinis) [inline sql/dd/types/index.h]:
```suggestion
  virtual void set_disabled(bool disable) noexcept = 0;
```

File: sql/dd/types/partition_index.h [hunk 8]
```diff
@@ -69,6 +69,14 @@ class Partition_index : virtual public Weak_object {
 
   const String_type &name() const { return index().name(); }
 
+  /////////////////////////////////////////////////////////////////////////
+  // disabled.
+  /////////////////////////////////////////////////////////////////////////
+
+  virtual bool is_disabled() const = 0;
```
Comment (laurynas-biveinis) [inline sql/dd/types/partition_index.h]:
```suggestion
  virtual bool is_disabled() const noexcept = 0;
```

File: sql/dd/types/partition_index.h [hunk 9]
```diff
@@ -69,6 +69,14 @@ class Partition_index : virtual public Weak_object {
...
+  virtual void set_disabled(bool disable) = 0;
```
Comment (laurynas-biveinis) [inline sql/dd/types/partition_index.h]:
```suggestion
  virtual void set_disabled(bool disable) noexcept = 0;
```

File: sql/dd/types/partition_index.h [hunk 10]
```diff
@@ -69,6 +69,14 @@ class Partition_index : virtual public Weak_object {
...
+  virtual Index::enum_index_type type() const = 0;
```
Comment (laurynas-biveinis) [inline sql/dd/types/partition_index.h]:
```suggestion
  virtual Index::enum_index_type type() const noexcept = 0;
```

File: sql/sql_table.cc [hunk 11]
```diff
@@ -13933,6 +13935,17 @@ static bool remove_secondary_keys(THD *thd, HA_CREATE_INFO *create_info,
       HA_ALTER_INPLACE_NOT_SUPPORTED)
     DBUG_RETURN(true);
 
+  for (auto index : *altered_table_def->indexes()) {
```
Comment (laurynas-biveinis) [inline sql/sql_table.cc]:
```suggestion
  for (const auto index : *altered_table_def->indexes()) {
```

File: sql/sql_table.cc [hunk 12]
```diff
@@ -13983,6 +13996,10 @@ static bool restore_secondary_keys(THD *thd, HA_CREATE_INFO *create_info,
   ha_alter_info.index_add_buffer =
       (uint *)thd->alloc(sizeof(uint) * alter_info->delayed_key_count);
 
+  for (auto index : *dd_disabled_sec_keys) {
```
Comment (laurynas-biveinis) [inline sql/sql_table.cc]:
```suggestion
  for (const auto index : *dd_disabled_sec_keys) {
```

File: storage/innobase/dict/dict0dd.cc [hunk 13]
```diff
@@ -1909,10 +1916,16 @@ void dd_write_table(dd::Object_id dd_space_id, Table *dd_table,
...
+    if (table->skip_alter_undo && dd_index->is_disabled()) {
+      ut_ad(dd_index->type() == dd::Index::IT_MULTIPLE);
+      continue;
+    }
+
     /* Don't assume the index orders are the same, even on
     CREATE TABLE. This could be called from TRUNCATE path,
     which would do some adjustment on FULLTEXT index, thus
     the out-of-sync order */
+
```
Comment (laurynas-biveinis) [inline storage/innobase/dict/dict0dd.cc]: Spurious

File: storage/innobase/handler/handler0alter.cc [hunk 14]
```diff
@@ -3807,7 +3807,16 @@ static void dd_commit_inplace_alter_table(
                                  FTS_DOC_ID_INDEX_NAME, col);
     }
 
-    dd_space_id = dd_first_index(old_dd_tab)->tablespace_id();
+    /* This can happen only with expanded fast index creation. On the
+    intermediate table during ALTER COPY, we drop secondary indexes using
+    inplace alter APIs. The old definition here is old copy of table. Hence we
+    should use new_dd_tab here for updating the dd::Indexes */
+    if (new_table->skip_alter_undo) {
```
Comment (laurynas-biveinis) [inline storage/innobase/handler/handler0alter.cc:3814]: Maybe
`dd_space_id = dd_first_index(new_table->skip_alter_undo ? new_dd_tab : old_dd_tab)->tablespace_id();`
?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
I like this approach very much.
My only "big" question is how it crash recovers for a failure while secondary index definitions are dropped? If any work is needed there, it can be filed as a follow-up issue

File: storage/innobase/handler/handler0alter.cc (hunk 14 above)
Comment (satya-bodapati) [reply storage/innobase/handler/handler0alter.cc:3814]: Dont like the readability of this suggestion.

**Comment by satya-bodapati** [comment]
Regarding the crash recovery, if there is crash after secondary indexes are removed, there is no loss. This is because, the entire ALTER is now one transaction and the secondary indexes removed are from the "newer version"/"interemdiate version" of table.

The original table remains intact. On crash, the original table is available. The question is whether this intermediate table is removed or not. I think it should be. The entire DD trx is rolled back.

Neverthelss, I will test this and if there is issue, will create a bug (follow-up)

File: storage/innobase/handler/handler0alter.cc (hunk 14 above)
Comment (laurynas-biveinis) [reply storage/innobase/handler/handler0alter.cc:3814]: Fair enough

**Comment by satya-bodapati** [comment]
Ready for review

**Review by laurynas-biveinis — APPROVED** [review]

**Comment by satya-bodapati** [comment]
https://ps.cd.percona.com/job/percona-server-8.0-param/210/

**Comment by satya-bodapati** [comment]
Recorded all_persisted_variables.test

**Pull request merged**